sdk/core/azure-core/azure/core/messaging.py

```diff
@@ -0,0 +1,148 @@
+    :vartype subject: str
+    :ivar specversion: Optional. The version of the CloudEvent spec. Defaults to "1.0"
+    :vartype specversion: str
+    :ivar id: An identifier for the event. The combination of id and source must be
+     unique for each distinct event. If not provided, a random UUID will be generated and used.
+    :vartype id: str
+    :ivar extensions: A CloudEvent MAY include any number of additional context attributes
+     with distinct names represented as key - value pairs. Each extension must be alphanumeric, lower cased
+     and must not exceed the length of 20 characters.
+    :vartype extensions: dict
+    """
+
+    def __init__(self, source, type, **kwargs):  # pylint: disable=redefined-builtin
+        # type: (str, str, **Any) -> None
+        self.source = source
+        self.type = type
+        self.specversion = kwargs.pop("specversion", "1.0")
+        self.id = kwargs.pop("id", str(uuid.uuid4()))
+        self.time = kwargs.pop("time", datetime.now(TZ_UTC))
+        self.datacontenttype = kwargs.pop("datacontenttype", None)
+        self.dataschema = kwargs.pop("dataschema", None)
+        self.subject = kwargs.pop("subject", None)
+        self.extensions = {}
+        _extensions = dict(kwargs.pop("extensions", {}))
+        for key in _extensions.keys():
+            if not key.islower() or not key.isalnum():
+                raise ValueError(
+                    "Extension attributes should be lower cased and alphanumeric."
+                )
+        self.extensions.update(_extensions)
+        self.data = kwargs.pop("data", None)
+
+    def __repr__(self):
+        return "CloudEvent(source={}, type={}, specversion={}, id={}, time={})".format(
+            self.source, self.type, self.specversion, self.id, self.time
+        )[:1024]
+
+    @classmethod
+    def from_dict(cls, event, **kwargs):
+        # type: (Dict, **Any) -> CloudEvent
+        """
+        Returns the deserialized CloudEvent object when a dict is provided.
+        :param event: The dict representation of the event which needs to be deserialized.
+        :type event: dict
+        :rtype: CloudEvent
+        """
+        reserved_attr = [
+            "data",
+            "data_base64",
+            "id",
+            "source",
+            "type",
+            "specversion",
+            "time",
+            "dataschema",
+            "datacontenttype",
+            "subject",
+        ]
+
+        data = event.get("data", None)
+        data_base64 = event.get("data_base64", None)
+        if data and data_base64:
+            raise ValueError(
+                "Invalid input. Only one of data and data_base64 must be present."
+            )
+
+        return cls(
+            id=event.get("id", None),
+            source=event.get("source", None),
+            type=event.get("type", None),
+            specversion=event.get("specversion", None),
+            data=data if data is not None else b64decode(data_base64),
```

Comment:
What is supposed to happen if the `event` has neither `"data"` nor `"data_base64"`?

Reply:
data will be none

Reply:
https://github.com/cloudevents/spec/blob/master/spec.md#event-data

data is optional, if neither data nor data_base64 is found in the incoming dict, we should not pass data at all

Reply:
yes - we are passing None when there is no data / data_base64 which is practically the same

Reply:
not the same :). Sending `data=null` and not sending data on the wire is different